decred/tests/unit/dcr/conftest.py

```diff
@@ -0,0 +1,40 @@
+"""
+Copyright (c) 2019, the Decred developers
+See LICENSE for details
+"""
+
+import pytest
+
+from decred.util import tinyhttp
+
+
+@pytest.fixture
+def http_get(request, monkeypatch):
+    """
+    Use this fixture in tests while defining the responses to be returned in
+    a "HTTP_GET_RESP" function-level dict with "uri" as keys.
+    """
+    get_resp = getattr(request.function, "HTTP_GET_RESP")
+
+    def mock_get(uri, **kwargs):
+        nonlocal get_resp
+        return get_resp[uri]
+
+    # Avoid calling the actual tinyhttp.get .
+    monkeypatch.setattr(tinyhttp, "get", mock_get)
+
+
+@pytest.fixture
+def http_post(request, monkeypatch):
+    """
+    Use this fixture in tests while defining the responses to be returned in
+    a "HTTP_POST_RESP" function-level dict with "(uri, repr(data))" as keys.
+    """
+    post_resp = getattr(request.function, "HTTP_POST_RESP")
+
+    def mock_post(uri, data, **kwargs):
+        nonlocal post_resp
+        return post_resp[(uri, repr(data))]
+
+    # Avoid calling the actual tinyhttp.post .
+    monkeypatch.setattr(tinyhttp, "post", mock_post)
```

Comment:
I want to suggest a slightly different pattern here. You probably know this, but the return value from the fixture is what is actually passed to the function. Instead of storing responses in a `dict` that is a property of the test function object, have the fixture return a function to queue up responses. 

```
@pytest.fixture
def http_post(monkeypatch):
    q = {}

    def mock_post(uri, data, **kwargs):
        return q[uri].pop()

    monkeypatch.setattr(tinyhttp, "post", mock_post)

    def queue(k, v):
        if not k in q:
            q[k] = []
        q[k].append(v)

    return queue
```

Allowing more than one response to be queued also enables more complicated paths to be tested.

Reply:
This is a brilliant idea! It is a bit more complex though, and I'm not sure what the advantage is (apart from the multiple responses per URI, that can be added to the current code is useful). Function attributes are unusual, true, but they are a documented feature of the language after all.

That aside, I'm a bit worried about the potential for interference among tests. Ideally each test would have a way to empty the queue, to avoid problems from some previous test leaving items in there.

Less importantly, you are showing the POST mock but using only the URI as the key, without the POST data. Does that mean that the latter may be left out of the key?

Reply:
> Function attributes are unusual

I agree. I also don't want to use introspection just because we can. It's a dirty business. 

> Ideally each test would have a way to empty the queue, to avoid problems with some previous test possibly leaving items in there.

I don't believe this is an issue. [Unless the `scope="module"` argument is passed](https://docs.pytest.org/en/latest/fixture.html#sharing-test-data) to the fixture decorator, the fixture will be invoked per-test, so every test will get it's own queue. 

There's actually an argument to be made for having it module-scoped. If the test leaves things in the queue, then it likely isn't doing what the programmer thinks it is. Catching that would be a good thing.

> Does that mean that the latter may be left out of the key?

My example was pretty bare-bones. Modify as needed.

Reply:
> the fixture will be invoked per-test, so every test will get it's own queue.

Indeed, my point was moot. Or maybe not? :wink:

> Catching that would be a good thing.

So you actually prefer the module-level scope. Ok, it works for me.